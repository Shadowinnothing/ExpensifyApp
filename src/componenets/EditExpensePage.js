import React from 'react';

const EditExpensePage = (props) => {

    return (
        <div>
            Editing Expense with ID of: {props.match.params.id}
        </div>
    );
};

export default EditExpensePage;