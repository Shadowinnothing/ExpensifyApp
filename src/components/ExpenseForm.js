import React, {Component} from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

import 'react-dates/lib/css/_datepicker.css'; // <- makes react-dates look sexy as hell

export default class ExpenseForm extends Component {

  constructor(props){
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '', // <- divide by 100 to convert pennies to number with a decimal
      createdAt: props.expense ? moment(props.expense.createdAt): moment(),
      calanderFocused: false,
      error: ''
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));
  };

  onNoteChange = (e) => {
    //const note = e.target.value;
    e.persist();
    this.setState(() => ({note: e.target.value}));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
      this.setState(() => ({amount}));
    }
  };

  onDateChange = (createdAt) => {
    if(createdAt){
      this.setState(() => ({createdAt}));
    }
  };

  onFocusChange = ({focused}) => {
    this.setState(() => ({calanderFocused: focused}));
  };

  onSubmit = (e) => {
    e.preventDefault(); // <- prevents full page refresh
    // error handler
    if(!this.state.description || !this.state.amount){
      this.setState(() => ({error: 'Please Provide a Description and Amount'}));
    } else {
      this.setState(() => ({error: ''}));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100, // <- multiply by 100 because app works in cents
        createdAt: this.state.createdAt.valueOf(), // <- valueof() is from moment documentation, makes date a UNIX timestamp
        note: this.state.note
      });
    }
  };

  render(){
    return (
      <div>

        {this.state.error && <p><b>{this.state.error}</b></p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />

          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />

          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calanderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={(day) => false}
          />

          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>

        <button>Add Expense</button>
        </form>

      </div>
    );
  };
}
