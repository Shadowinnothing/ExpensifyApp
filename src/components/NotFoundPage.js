import React from 'react';

import {Link} from 'react-router-dom';

// Used for 404 page, plays youtube video
import YouTube from 'react-youtube';

const _404Opts = {
    height: '390', 
    width: '640',
    playerVars: {
        autoplay: 1
    }
}
const NotFoundPage = () => (
    <div>
        <h1>404 Not Found</h1>

        <YouTube 
            videoId="ZZ5LpwO-An4"
            opts={_404Opts}
        />

        <br />
        <Link to="/">Go Home</Link>
    </div>
);

export default NotFoundPage;