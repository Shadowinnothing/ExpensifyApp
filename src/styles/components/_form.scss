.form {
  display: flex;
  flex-direction: column;

  /* For every child in the form */
  > * {
    margin-bottom: $m-size;
  }
};

.form__error {
  margin: 0 0 $m-size 0;
  font-style: italic;
};
