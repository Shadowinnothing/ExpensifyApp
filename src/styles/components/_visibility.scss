.show-for-mobile {
  @media (min-width: $desktop-breakpoint) {
    display: none;
  }
};

.show-for-desktop {
  /* Subtract .01rem so neither CSS script runs at the same time */
  @media (max-width: $desktop-breakpoint - .01rem) {
    display: none;
  }
};
