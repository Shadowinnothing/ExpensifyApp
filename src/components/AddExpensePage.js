import React, {Component} from 'react';
import {connect} from 'react-redux';

import ExpenseForm from './ExpenseForm';
import {startAddExpense} from '../actions/expenses';

export class AddExpensePage extends Component {
  onSubmit = (expense) => {
    //props.dispatch(addExpense(expense));
    this.props.startAddExpense(expense);
    this.props.history.push('/') // <- switch over to dashboard without full page reload
  };
  render(){
    return (
      <div>
          <h1>Add Expense</h1>
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
      </div>
    );
  };
};

// similar to mapStatesToProp but works with dispatch on the store
// returns an object props that make a call to dispatch
// does the same thing as the first commented line in AddExpensePage, but makes it easier to test
const mapDispatchToProps = (dispatch) => ({startAddExpense: (expense) => dispatch(startAddExpense(expense)) });

// mapStatesToProp is unedined
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
