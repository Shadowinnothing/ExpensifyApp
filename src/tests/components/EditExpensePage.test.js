import React from 'react';
import {shallow} from 'enzyme';

import {EditExpensePage} from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

describe('components/EditExpensePage', () => {

  let wrapper, editExpenseSpy, removeExpenseSpy, historySpy;

  beforeEach(() => {
    editExpenseSpy = jest.fn();
    removeExpenseSpy = jest.fn();
    historySpy = {push: jest.fn()}
    wrapper = shallow(<EditExpensePage
      expense={expenses[0]}
      editExpense={editExpenseSpy}
      removeExpense={removeExpenseSpy}
      history={historySpy}
    />);

  });

  test('should correctly render EditExpensePage', () => {
    expect(wrapper).toMatchSnapshot();
  })

  test('should handle editExpenseSpy', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
    expect(editExpenseSpy).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
  });

  test('should handle removeExpenseSpy', () => {
    wrapper.find('button').simulate('click');
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
    expect(removeExpenseSpy).toHaveBeenLastCalledWith({id: expenses[0].id});
  });

});
