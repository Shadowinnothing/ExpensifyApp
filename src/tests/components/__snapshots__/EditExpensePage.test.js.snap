// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/EditExpensePage should correctly render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 350,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
